File: js/scripts/detect-code-changes.mjs
#!/usr/bin/env node

/**
 * Detect code changes for CI/CD pipeline
 *
 * This script detects what types of files have changed between two commits
 * and outputs the results for use in GitHub Actions workflow conditions.
 *
 * Key behavior:
 * - For PRs: compares PR head against base branch
 * - For pushes: compares HEAD against HEAD^
 * - Excludes certain folders and file types from "code changes" detection
 *
 * Excluded from code changes (don't require changesets):
 * - Markdown files (*.md) in any folder
 * - .changeset/ folder (changeset metadata)
 * - docs/ folder (documentation)
 * - experiments/ folder (experimental scripts)
 * - examples/ folder (example scripts)
 *
 * Usage:
 *   node scripts/detect-code-changes.mjs
 *
 * Environment variables (set by GitHub Actions):
 *   - GITHUB_EVENT_NAME: 'pull_request' or 'push'
 *   - GITHUB_BASE_SHA: Base commit SHA for PR
 *   - GITHUB_HEAD_SHA: Head commit SHA for PR
 *
 * Outputs (written to GITHUB_OUTPUT):
 *   - mjs-changed: 'true' if any .mjs files changed
 *   - js-changed: 'true' if any .js files changed
 *   - package-changed: 'true' if package.json changed
 *   - docs-changed: 'true' if any .md files changed
 *   - workflow-changed: 'true' if any .github/workflows/ files changed
 *   - any-code-changed: 'true' if any code files changed (excludes docs, changesets, experiments, examples)
 */

import { execSync } from 'child_process';
import { appendFileSync } from 'fs';
import { pathToFileURL } from 'url';

/**
 * Execute a shell command and return trimmed output
 * @param {string} command - The command to execute
 * @returns {string} - The trimmed command output
 */
function exec(command) {
  try {
    return execSync(command, { encoding: 'utf-8' }).trim();
  } catch (error) {
    console.error(`Error executing command: ${command}`);
    console.error(error.message);
    return '';
  }
}

/**
 * Write output to GitHub Actions output file
 * @param {string} name - Output name
 * @param {string} value - Output value
 */
function setOutput(name, value) {
  const outputFile = process.env.GITHUB_OUTPUT;
  if (outputFile) {
    appendFileSync(outputFile, `${name}=${value}\n`);
  }
  console.log(`${name}=${value}`);
}

/**
 * Get the list of changed files between two commits
 * @returns {string[]} Array of changed file paths
 */
export function getChangedFiles() {
  const eventName = process.env.GITHUB_EVENT_NAME || 'local';

  if (eventName === 'pull_request') {
    const baseSha = process.env.GITHUB_BASE_SHA;
    const headSha = process.env.GITHUB_HEAD_SHA;

    if (baseSha && headSha) {
      console.log(`Comparing PR: ${baseSha}...${headSha}`);
      try {
        // Ensure we have the base commit
        try {
          execSync(`git cat-file -e ${baseSha}`, { stdio: 'ignore' });
        } catch {
          console.log('Base commit not available locally, attempting fetch...');
          execSync(`git fetch origin ${baseSha}`, { stdio: 'inherit' });
        }
        const output = exec(`git diff --name-only ${baseSha} ${headSha}`);
        return output ? output.split('\n').filter(Boolean) : [];
      } catch (error) {
        console.error(`Git diff failed: ${error.message}`);
      }
    }
  }

  // For push events or fallback
  console.log('Comparing HEAD^ to HEAD');
  try {
    const output = exec('git diff --name-only HEAD^ HEAD');
    return output ? output.split('\n').filter(Boolean) : [];
  } catch {
    // If HEAD^ doesn't exist (first commit), list all files in HEAD
    console.log('HEAD^ not available, listing all files in HEAD');
    const output = exec('git ls-tree --name-only -r HEAD');
    return output ? output.split('\n').filter(Boolean) : [];
  }
}

/**
 * Path of the current working directory relative to the repository root.
 *
 * `git diff --name-only` always prints paths relative to the **repository
 * root**, but this script runs with `working-directory: ./js` and compares
 * those paths against package-relative prefixes such as `examples/`. In this
 * monorepo the real paths are `js/examples/...`, so before issue #41 none of
 * the exclusions ever matched and `package.json` could never be detected -- the
 * same class of defect as issue #39. Resolving the prefix at run time keeps the
 * script correct both here and in a single-package checkout, where the prefix
 * is empty.
 *
 * @returns {string} The prefix, ending with `/`, or `''` at the repository root
 */
export function getPathPrefix() {
  return exec('git rev-parse --show-prefix');
}

/**
 * Re-express repository-root-relative paths as package-relative paths,
 * dropping everything that lives outside this package.
 *
 * @param {string[]} changedFiles - Repository-root-relative paths
 * @param {string} prefix - Package prefix such as `js/`
 * @returns {string[]} Package-relative paths
 */
export function toPackagePaths(changedFiles, prefix) {
  if (!prefix) {
    return changedFiles;
  }
  return changedFiles
    .filter((file) => file.startsWith(prefix))
    .map((file) => file.slice(prefix.length));
}

/**
 * Check whether a repository-root-relative path is a workflow definition.
 *
 * @param {string} filePath - Repository-root-relative path
 * @returns {boolean} True for files under `.github/workflows/`
 */
export function isWorkflowFile(filePath) {
  return filePath.startsWith('.github/workflows/');
}

/**
 * Check if a file should be excluded from code changes detection
 * @param {string} filePath - The file path to check
 * @returns {boolean} True if the file should be excluded
 */
export function isExcludedFromCodeChanges(filePath) {
  // Exclude markdown files in any folder
  if (filePath.endsWith('.md')) {
    return true;
  }

  // Exclude specific folders from code changes
  const excludedFolders = ['.changeset/', 'docs/', 'experiments/', 'examples/'];

  for (const folder of excludedFolders) {
    if (filePath.startsWith(folder)) {
      return true;
    }
  }

  return false;
}

/**
 * Classify a set of repository-root-relative changed paths.
 *
 * Split out from `detectChanges` so the classification can be unit tested
 * without a git repository.
 *
 * @param {string[]} changedFiles - Repository-root-relative changed paths
 * @param {string} prefix - Package prefix such as `js/`
 * @returns {{outputs: Record<string, string>, packageFiles: string[], codeChangedFiles: string[]}}
 */
export function classifyChanges(changedFiles, prefix) {
  const packageFiles = toPackagePaths(changedFiles, prefix);
  const workflowFiles = changedFiles.filter(isWorkflowFile);

  // Code changes are judged on package-relative paths, so the documented
  // `examples/`, `experiments/`, `docs/` and `.changeset/` exclusions apply.
  const codeChangedFiles = packageFiles.filter(
    (file) => !isExcludedFromCodeChanges(file)
  );

  // A workflow change still counts as a code change: it can alter how this
  // package is built and published even when no package file moved.
  const codePattern = /\.(mjs|js|json|yml|yaml)$/;
  const codeChanged =
    codeChangedFiles.some((file) => codePattern.test(file)) ||
    workflowFiles.length > 0;

  return {
    packageFiles,
    codeChangedFiles,
    outputs: {
      'mjs-changed': String(packageFiles.some((f) => f.endsWith('.mjs'))),
      'js-changed': String(packageFiles.some((f) => f.endsWith('.js'))),
      'package-changed': String(packageFiles.includes('package.json')),
      'docs-changed': String(packageFiles.some((f) => f.endsWith('.md'))),
      'workflow-changed': String(workflowFiles.length > 0),
      'any-code-changed': String(codeChanged),
    },
  };
}

/**
 * Main function to detect changes
 */
export function detectChanges() {
  console.log('Detecting file changes for CI/CD...\n');

  const changedFiles = getChangedFiles();
  const prefix = getPathPrefix();

  console.log(`Package prefix: ${prefix || '(repository root)'}`);
  console.log('Changed files:');
  if (changedFiles.length === 0) {
    console.log('  (none)');
  } else {
    changedFiles.forEach((file) => console.log(`  ${file}`));
  }
  console.log('');

  const { codeChangedFiles, outputs } = classifyChanges(changedFiles, prefix);

  console.log('Files considered as code changes:');
  if (codeChangedFiles.length === 0) {
    console.log('  (none)');
  } else {
    codeChangedFiles.forEach((file) => console.log(`  ${file}`));
  }
  console.log('');

  for (const [name, value] of Object.entries(outputs)) {
    setOutput(name, value);
  }

  console.log('\nChange detection completed.');
}

// Run the detection unless this module was imported by a test.
if (
  process.argv[1] &&
  import.meta.url === pathToFileURL(process.argv[1]).href
) {
  detectChanges();
}

File: python/scripts/detect_code_changes.py
#!/usr/bin/env python3
"""Detect code changes for the Python CI/CD pipeline.

This script decides whether a push or pull request touched anything that needs
the expensive jobs to run. It mirrors ``scripts/detect_code_changes.py`` in
``link-foundation/python-ai-driven-development-pipeline-template`` and the
JavaScript, Rust and C# siblings in this repository.

Key behaviour:

- For pull requests: compares the pull request head against its base.
- For pushes: compares ``HEAD`` against ``HEAD^``.
- Paths are judged *relative to this package*. ``git diff --name-only`` prints
  repository-root paths, so in this monorepo the real paths are
  ``python/examples/demo.py``; without stripping the ``python/`` prefix the
  documented exclusions below would never match. That mistake is what issue #39
  was about, and the JavaScript and Rust scripts carried it until issue #41.

Excluded from code changes (they never need a changelog fragment):

- Markdown files (``*.md``) anywhere.
- ``changelog.d/`` (changelog fragments)
- ``docs/`` (documentation)
- ``experiments/`` (experimental scripts)
- ``examples/`` (example scripts)

Usage::

    python scripts/detect_code_changes.py

Environment variables (set by GitHub Actions):

- ``GITHUB_EVENT_NAME``: ``pull_request`` or ``push``
- ``GITHUB_BASE_SHA``: base commit SHA for a pull request
- ``GITHUB_HEAD_SHA``: head commit SHA for a pull request

Outputs (appended to ``GITHUB_OUTPUT``):

- ``py-changed``: any ``.py`` file in this package changed
- ``toml-changed``: any ``.toml`` file in this package changed
- ``docs-changed``: any ``.md`` file in this package changed
- ``workflow-changed``: any ``.github/workflows/`` file changed
- ``any-code-changed``: any code file changed, exclusions applied
"""

from __future__ import annotations

import os
import subprocess

EXCLUDED_FOLDERS = (
    "changelog.d/",
    "docs/",
    "examples/",
    "experiments/",
)
CODE_EXTENSIONS = (".py", ".toml", ".yml", ".yaml", ".cfg")
WORKFLOW_PREFIX = ".github/workflows/"


def exec_command(command: list[str]) -> str:
    """Run a command and return its trimmed stdout, or an empty string."""
    try:
        result = subprocess.run(  # noqa: S603 - fixed argument lists only
            command,
            capture_output=True,
            text=True,
            check=True,
        )
    except (OSError, subprocess.CalledProcessError) as error:
        print(f"Error executing command: {' '.join(command)}")
        print(error)
        return ""
    return result.stdout.strip()


def set_output(name: str, value: str) -> None:
    """Append a GitHub Actions job output and echo it into the log."""
    output_file = os.environ.get("GITHUB_OUTPUT")
    if output_file:
        with open(output_file, "a", encoding="utf-8") as handle:
            handle.write(f"{name}={value}\n")
    print(f"{name}={value}")


def get_path_prefix() -> str:
    """Return the working directory relative to the repository root.

    ``python/`` in this monorepo, and ``""`` in a single-package checkout, where
    every path is already package-relative.
    """
    return exec_command(["git", "rev-parse", "--show-prefix"])


def get_changed_files() -> list[str]:
    """Return the repository-root-relative paths changed by this event."""
    event_name = os.environ.get("GITHUB_EVENT_NAME", "local")

    if event_name == "pull_request":
        base_sha = os.environ.get("GITHUB_BASE_SHA")
        head_sha = os.environ.get("GITHUB_HEAD_SHA")
        if base_sha and head_sha:
            print(f"Comparing PR: {base_sha}...{head_sha}")
            if not exec_command(["git", "cat-file", "-t", base_sha]):
                print("Base commit not available locally, attempting fetch...")
                exec_command(["git", "fetch", "origin", base_sha])
            output = exec_command(["git", "diff", "--name-only", base_sha, head_sha])
            return [f for f in output.split("\n") if f]

    print("Comparing HEAD^ to HEAD")
    output = exec_command(["git", "diff", "--name-only", "HEAD^", "HEAD"])
    if output:
        return [f for f in output.split("\n") if f]

    # First commit in the repository: there is no HEAD^ to compare against.
    print("HEAD^ not available, listing all files in HEAD")
    output = exec_command(["git", "ls-tree", "--name-only", "-r", "HEAD"])
    return [f for f in output.split("\n") if f]


def to_package_paths(changed_files: list[str], prefix: str) -> list[str]:
    """Re-express root-relative paths as package-relative, dropping the rest."""
    if not prefix:
        return list(changed_files)
    return [f.removeprefix(prefix) for f in changed_files if f.startswith(prefix)]


def is_workflow_file(file_path: str) -> bool:
    """Return True for a repository-root workflow definition."""
    return file_path.startswith(WORKFLOW_PREFIX)


def is_excluded_from_code_changes(file_path: str) -> bool:
    """Return True when a *package-relative* path is not a code change."""
    if file_path.endswith(".md"):
        return True
    return file_path.startswith(EXCLUDED_FOLDERS)


def classify_changes(changed_files: list[str], prefix: str) -> tuple[dict[str, str], list[str]]:
    """Classify changed paths, returning the outputs and the code files."""
    package_files = to_package_paths(changed_files, prefix)
    workflow_files = [f for f in changed_files if is_workflow_file(f)]

    code_changed_files = [f for f in package_files if not is_excluded_from_code_changes(f)]
    # A workflow change counts as a code change: it can alter how this package
    # is built and published even when no package file moved.
    code_changed = bool(workflow_files) or any(
        f.endswith(CODE_EXTENSIONS) for f in code_changed_files
    )

    outputs = {
        "py-changed": str(any(f.endswith(".py") for f in package_files)).lower(),
        "toml-changed": str(any(f.endswith(".toml") for f in package_files)).lower(),
        "docs-changed": str(any(f.endswith(".md") for f in package_files)).lower(),
        "workflow-changed": str(bool(workflow_files)).lower(),
        "any-code-changed": str(code_changed).lower(),
    }
    return outputs, code_changed_files


def detect_changes() -> None:
    """Detect the changes of the current event and publish them as outputs."""
    print("Detecting file changes for CI/CD...\n")

    changed_files = get_changed_files()
    prefix = get_path_prefix()

    print(f"Package prefix: {prefix or '(repository root)'}")
    print("Changed files:")
    if not changed_files:
        print("  (none)")
    else:
        for file in changed_files:
            print(f"  {file}")
    print()

    outputs, code_changed_files = classify_changes(changed_files, prefix)

    print("Files considered as code changes:")
    if not code_changed_files:
        print("  (none)")
    else:
        for file in code_changed_files:
            print(f"  {file}")
    print()

    for name, value in outputs.items():
        set_output(name, value)

    print("\nChange detection completed.")


if __name__ == "__main__":
    detect_changes()
